README.md
Pktgen - Traffic Generator powered by DPDK
=====================================================

**Pktgen is a traffic generator powered by DPDK at wire rate traffic with 64 byte frames.**

** (Pktgen) Sounds like 'Packet-Gen'**

---
```
**Copyright &copy; \<2010-2024\>, Intel Corporation. All rights reserved.**

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

- Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.

- Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in
  the documentation and/or other materials provided with the
  distribution.

- Neither the name of Intel Corporation nor the names of its
  contributors may be used to endorse or promote products derived
  from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
OF THE POSSIBILITY OF SUCH DAMAGE.

SPDX-License-Identifier: BSD-3-Clause

Pktgen: Created 2010-2024 by Keith Wiles @ Intel.com
```
---
Pktgen documentation can be found here: https://pktgen.github.io/Pktgen-DPDK/

---
Using the tools/run.py script to setup and run pktgen with different configurations. The configuration files are located in the cfg directory with filenames ending in .cfg.

To use a configuration file;
```
$ ./tools/run.py -s default  # to setup the ports and attach them to DPDK (only needed once per boot)

$ ./tools/run.py default     # Run the default configuration
```
The configuration files are python scripts or a set of variables that run.py uses to initialize and run pktgen.
Here is an example of the default.cfg file:

```
	description = 'A Pktgen default simple configuration'

	# Setup configuration
	setup = {
	    'exec': (
		'sudo', '-E'
		),

	    'devices': (
		    '81:00.0', '81:00.1', '81:00.2', '81:00.3',
		    '83:00.0', '83:00.1', '83:00.2', '83:00.3'
		    ),
	    # UIO module type, igb_uio, vfio-pci or uio_pci_generic
	    'uio': 'vfio-pci'
	    }

	# Run command and options
	run = {
	    'exec': (
		'sudo', '-E'
		),

	    # Application name and use app_path to help locate the app
	    'app_name': 'pktgen',

	    # using (sdk) or (target) for specific variables
	    # add (app_name) of the application
	    # Each path is tested for the application
	    'app_path': (
		'./app/%(target)s/%(app_name)s',
		'%(sdk)s/%(target)s/app/%(app_name)s',
		),

	    'cores': '14,15-22',
	    'nrank': '4',
	    'proc': 'auto',
	    'log': '7',
	    'prefix': 'pg',

	    'blocklist': (
		#'81:00.0', '81:00.1', '81:00.2', '81:00.3',
		#'83:00.0', '83:00.1', '83:00.2', '83:00.3',
		'81:00.2', '81:00.3',
		'83:00.2', '83:00.3'
		),

	    'opts': (
		'-T',
		'-P',
		),
	    'map': (
		'[15:16].0',
		'[17:18].1',
		'[19:20].2',
		'[21:22].3'
		),

	    'theme': 'themes/black-yellow.theme'
	}
```

------------------------------------------------------------------------------------------
-- Two Pktgens running on the same machine with connected via a loopback ports

Look at the two new files pktgen-1.cfg and pktgen-2.cfg in the cfg directory for some help on
the configuration to run two pktgens at the same time on the same machine.

------------------------------------------------------------------------------------------
-- Socket Support for Pktgen.

Pktgen has a TCP socket connection to allow you to control Pktgen from a remote
program or console. The TCP connection is using port 0x5606 or 22022 and presents
a Lua command shell interface. If you telnet to the machine running Pktgen on port
22022 you will get a lua command shell like interface. This interface does not have
a command line prompt, but you can issue Lua code or load script files from the local
disk of the machine. You can also send programs to the remote Pktgen machine to
load scripts from a remote location.

One method to connect to Pktgen is using telnet, but another method would be to
use 'socat' program on a Linux machine. The socat program is very powerfull application
and can do a lot of things. I used socat to debug Pktgen using the following
command, which gives me a readline inteface to Pktgen's socket interface.
Remember to use the '-G' option or -g host:port pktgen option, then make sure you
use the same address in the socat line.

```
$ socat -d -d READLINE TCP4:localhost:22022
```

'You will see socat create the connection and then wait for Lua command scripts for you'
To exit this command type Control-D to exit and close the connection.

You can also just send Pktgen a script file and display the ouptut.

---------
    $ socat - TCP4:localhost:22022 < test/hello-world.lua

    Lua Version: Lua 5.3
    Pktgen Version : 3.6.1
    Pktgen Copyright : Copyright(c) `<2010-2024>`, Intel Corporation
    Pktgen Authors : Keith Wiles @ Intel Corporation

Hello World!!!!
--------

Here is the program I sent Pktgen:

    $ cat test/hello-world.lua
    package.path = package.path ..";?.lua;test/?.lua;app/?.lua;"

    printf("Lua Vesrion: %s\n", pktgen.info.Lua_Version);
    printf("Pktgen Version : %s\n", pktgen.info.Pktgen_Version);
    printf("Pktgen Copyright : %s\n", pktgen.info.Pktgen_Copyright);
    printf("Pktgen Authors : %s\n", pktgen.info.Pktgen_Authors);

    printf("\nHello World!!!!\n");
-----------

Here is a command from my Mac Book pro laptop, which loads a file from the local
disk where Pktgen is running and then we execute the file with 'f()'.

------------------

    $ socat READLINE TCP4:172.25.40.163:22022
    f,e = loadfile("test/hello-world.lua")
    f()
    Lua Version: Lua 5.3
    Pktgen Version : 3.6.1
    Pktgen Copyright : Copyright(c) `<2010-2024>`, Intel Corporation
    Pktgen Authors : Keith Wiles @ Intel Corporation

    Hello World!!!!
    <Control-D>
------------------

You can also just send it commands via echo.

-----------------

    $ echo "f,e = loadfile('test/hello-world.lua'); f();"| socat - TCP4:172.25.40.163:22022
    Lua Version: Lua 5.3
    Pktgen Version : 3.6.1
    Pktgen Copyright : Copyright(c) `<2010-2024>`, Intel Corporation
    Pktgen Authors : Keith Wiles @ Intel Corporation

    Hello World!!!!
----------------------

Keith Wiles @ Intel Corporation


changelog.txt
Pktgen-DPDK - Traffic Generator powered by DPDK
=====================================================

**Pktgen-DPDK is a traffic generator powered by DPDK at wire rate traffic with 64 byte frames.**

** (Pktgen) Sounds like 'Packet-Gen'**

**=== Modifications ===**
 - 23.11.1 - Update and fix performance issues.
 - 23.11.0 - Add a new application pktperf to test performance improvements and validation.
 - 23.10.2 - Fix range tcp flags and help messages
 - 23.10.1 - Fix rte_thread_setname() change to rte_thread_set_name() for lua builds
 - 23.10.0 - Change TX rate calculations to be more accurate
             Change timer routines to use DPDK timer manager
             Minor cleanup and bump version number.
 - 23.06.1 - cleanup the code with clang-format.
 - 23.06.0 - Fix up problem using "all" with portlist routines.
 - 23.03.1 - In latency mode avg_cycles was only calculated when not using Lua.
 - 23.03.0 - Update and fix a number of issues with latency. Some values were not updated
             or displayed on screen or in a Lua script when getting port stats.
             Latency packets are not able to be sent in any mode, instead of only being able
             to send and report latencies in latency mode.
             When using a portlist in lua it can be a single port in a string, a string range
             or a single port using a Lua Number or Integer
 - 22.07.2 - Changed the number of queues from 16 to 64.
 - 22.07.1 - Update copyright dates and other minor fixes
 - 22.07.0 - Add new rx/tx burst options.
             Fix GUI build and stats.
             Make build with DPDK more compatible.
             Add max latency reporting.
             Allow building on RISC-V platforms.
             Add new rfc2455 script
             Increase the number of mbufs allocated to support better performance.
             Add restart command to try and cleanup the DPDK PMD or mbufs a start RX/TX.
             Update documentation contribution guidlines.
             Remove the -j option.
             Add relative delay to compute rate stats and support multple queue stats.
 - 22.04.1 - Fix Lua build issues around new stats structure changes.
 - 22.02.0 - Remove the backward compatability builds with older versions of DPDK.
             It has become too difficult to maintain building with older versions and I believe
             the older DPDK version builds did not work as expected. Please only build DPDK with
             the lastest version of DPDK or at least the previous version, but I do not guarantee
             it will work. I will try to keep up with the latest release version of DPDK.
 - 21.11.0 - Fix DPDK defines being renamed of deprecated. Also rte_eth_devices was made static.
 - 21.10.0 - Fix up ARP processing, change pcap files to rw-rw-rw mode, pcap capture on the correct
             numa node, increase the TX burst size, update to work with 21.11 DPDK, update the
             tx_failed counter, add TCP flags/seq/ack, pcap use max packet size for mempool, pcap
             is a multiple of pcap length, fixup AVX/AVX2 support, ipv6 updates.
 - 21.03.1 - Fixup AVX and AVX2 compiler support, if enabled in meson_options.txt file.
 - 21.03.0 - Fixup DPDK name changes and other minor problems.
 - 21.02.0 - Cleanup display to be more compact and fix up colors
             Update the copyright notice years
             Fix the pcap issues with allocating memory on the wrong socket.
             Add thousand separators to see rates and values easier.
 - 21.01.2 - Add updated IPv6 support.
 - 21.01.1 - Update the version number with fix for sending special mbufs for input processing.
 - 21.01.0 - fix warning message spit out by DPDK and add lua-5.4 support in build.
 - 20.12.0 - merge in latency test support with jitter and a number of minor cleanups
 - 20.11.3 - fix changes in DPDK 20.11-rc4
 - 20.11.2 - The dynfield0 was removed from rte_mbuf header
 - 20.11.1 - Fix net_bond library name change for older DPDK versions.
 - 20.11.0 - Fixup build to DPDK PMD name changes (bonding driver)
             Fix the macro and function API name changes in DPDK 20.11
 - 20.10.1 - Add option to disable avx2 support.
 - 20.10.0 - fix ENA memory allocation and add setting rx/tx descriptor rings from cmdline.
 - 20.09.0 - Fix the clearing of the port stats.
             Fix the lua range commands to be a table only. Use pktgen.range.XYZ() to set range values.
             Removed the use of pktgen.XYZ() type commands and require the range table to be used.
 - 20.08.0 - Add better latency support with saving the results to a file.
             Updated the INSTALL.md to help building DPDK and Pktgen.
             Number of minor bugs or cleanup.
 - 20.07.0 - Fix lua/gui builds to find the lua lib/headers on ubuntu 20.04
             Update the INSTALL text to locate DPDK pkg-config, this also fixes Lua (I hope in all cases)
 - 20.06.0 - Convert to using meson/ninja only builds.
             Fix the default mbuf size.
             Fix sending of special or ARP packets.
             Add Makefile to use tools/pktgen-build.sh script for meson/ninja
             Fix up building the docs using meson/ninja.
             New target directory for install is $PKTGEN_DIR/usr/local, need to build and
             install DPDK using meson/ninja.
 - 20.04.0 - Disable Lua by default in meson_options.tst file for meson/ninja builds.
             Lua support is disabled by default now as we move to meson/ninja builds only.
             When building with meson/ninja DPDK needs to be built meson/ninja and installed.
 - 20.03.0 - Fix a typo in run.py and add af_xdp config file.
 - 20.02.0 - Fix link variable to be lk to not conflect with global link struct.
             fix generating sequences with vlan header
             fix sequence help message
 - 19.12.0 - Add per queue stats when the hardware does not support it. Use
             the 'page stats' command to see data and use 'port X' to switch
	     ports.

app/pktgen.h
/*-
 * Copyright(c) <2010-2023>, Intel Corporation. All rights reserved.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

/* Created 2010 by Keith Wiles @ intel.com */

#ifndef _PKTGEN_H_
#define _PKTGEN_H_

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <sys/types.h>
#include <unistd.h>
#include <string.h>
#include <sys/queue.h>
#include <netinet/in.h>
#include <net/if.h>
#include <fcntl.h>
#include <setjmp.h>
#include <stdarg.h>
#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#include <termios.h>
#include <sys/ioctl.h>
#include <libgen.h>
#include <linux/if_tun.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <poll.h>
#include <assert.h>
#include <time.h>

#include <pg_compat.h>
#include <rte_config.h>

#include <rte_errno.h>
#include <rte_log.h>
#include <rte_tailq.h>
#include <rte_common.h>
#include <rte_memory.h>
#include <rte_memcpy.h>
#include <rte_memzone.h>
#include <rte_malloc.h>
#include <rte_eal.h>
#include <rte_per_lcore.h>
#include <rte_launch.h>
#include <rte_atomic.h>
#include <rte_cycles.h>
#include <rte_prefetch.h>
#include <rte_lcore.h>
#include <rte_branch_prediction.h>
#include <rte_pci.h>
#include <rte_random.h>
#include <rte_timer.h>
#include <rte_ether.h>
#include <rte_ethdev.h>
#include <rte_ring.h>
#include <rte_mempool.h>
#include <rte_mbuf.h>
#include <rte_ip.h>
#include <rte_udp.h>
#include <rte_tcp.h>
#include <rte_dev.h>
#include <rte_time.h>

#include <copyright_info.h>
#include <l2p.h>
#include <port_config.h>
#include <core_info.h>

#include <pg_inet.h>
#include <cksum.h>

#include <mbuf.h>
#include <coremap.h>
#include <lscpu.h>
#include <utils.h>

#ifdef LUA_ENABLED
#include "lua_config.h"
#include "lauxlib.h"
#endif

#include "pktgen-port-cfg.h"
#include "pktgen-capture.h"
#include "pktgen-log.h"
#include "pktgen-latency.h"
#include "pktgen-random.h"
#include "pktgen-rate.h"
#include "pktgen-seq.h"
#include "pktgen-version.h"

#include <cli.h>

#ifdef LUA_ENABLED
#include <lua_config.h>
#endif

#ifdef __cplusplus
extern "C" {
#endif

#define MAX_MATRIX_ENTRIES 128
#define MAX_STRING         256
#define Million            (uint64_t)(1000000UL)
#define Billion            (uint64_t)(1000000000UL)

#define iBitsTotal(_x) (uint64_t)(((_x.ipackets * PKT_OVERHEAD_SIZE) + _x.ibytes) * 8)
#define oBitsTotal(_x) (uint64_t)(((_x.opackets * PKT_OVERHEAD_SIZE) + _x.obytes) * 8)

#define _do(_exp) \
    do {          \
        _exp;     \
    } while ((0))

#ifndef RTE_ETH_FOREACH_DEV
#define RTE_ETH_FOREACH_DEV(p) for (_p = 0; _p < pktgen.nb_ports; _p++)
#endif

#define forall_ports(_action)           \
    do {                                \
        uint16_t pid;                   \
                                        \
        RTE_ETH_FOREACH_DEV(pid)        \
        {                               \
            port_info_t *pinfo;         \
                                        \
            info = &pktgen.info[pid];   \
            if (pinfo->seq_pkt == NULL) \
                continue;               \
            _action;                    \
        }                               \
    } while ((0))

#define foreach_port(_portlist, _action)                  \
    do {                                                  \
        uint64_t *_pl = (uint64_t *)&_portlist;           \
        uint16_t pid, idx, bit;                           \
                                                          \
        RTE_ETH_FOREACH_DEV(pid)                          \
        {                                                 \
            port_info_t *pinfo;                           \
                                                          \
            idx = (pid / (sizeof(uint64_t) * 8));         \
            bit = (pid - (idx * (sizeof(uint64_t) * 8))); \
            if ((_pl[idx] & (1LL << bit)) == 0)           \
                continue;                                 \
            pinfo = l2p_get_port_pinfo(pid);              \
            if (pinfo->seq_pkt == NULL)                   \
                continue;                                 \
            _action;                                      \
        }                                                 \
    } while ((0))

typedef enum {
    PACKET_CONSUMED = 0,
    UNKNOWN_PACKET  = 0xEEEE,
    DROP_PACKET     = 0xFFFE,
    FREE_PACKET     = 0xFFFF
} pktType_e;

enum {
    MAX_SCRN_ROWS = 44,
    MAX_SCRN_COLS = 132,

    COLUMN_WIDTH_0 = 22,
    COLUMN_WIDTH_1 = 24,
    COLUMN_WIDTH_3 = 24,

    /* Row locations for start of data */
    PAGE_TITLE_ROWS = 1,
    PORT_FLAGS_ROWS = 1,
    LINK_STATE_ROWS = 1,
    PKT_RATE_ROWS   = 9,
    PKT_SIZE_ROWS   = 10,
    PKT_TOTALS_ROWS = 7,
    IP_ADDR_ROWS    = 12,

    PAGE_TITLE_ROW = 1,
    PORT_FLAGS_ROW = (PAGE_TITLE_ROW + PAGE_TITLE_ROWS),
    LINK_STATE_ROW = (PORT_FLAGS_ROW + PORT_FLAGS_ROWS),
    PKT_RATE_ROW   = (LINK_STATE_ROW + LINK_STATE_ROWS),
    PKT_SIZE_ROW   = (PKT_RATE_ROW + PKT_RATE_ROWS),
    PKT_TOTALS_ROW = (PKT_SIZE_ROW + PKT_SIZE_ROWS),
    IP_ADDR_ROW    = (PKT_TOTALS_ROW + PKT_TOTALS_ROWS),

    DEFAULT_NETMASK        = 0xFFFFFF00,
    DEFAULT_IP_ADDR        = (192 << 24) | (168 << 16),
    DEFAULT_TX_COUNT       = 0, /* Forever */
    DEFAULT_TX_RATE        = 100,
    DEFAULT_PRIME_COUNT    = 1,
    DEFAULT_SRC_PORT       = 1234,
    DEFAULT_DST_PORT       = 5678,
    DEFAULT_TTL            = 64,
    DEFAULT_TCP_SEQ_NUMBER = 0x12378,
    MAX_TCP_SEQ_NUMBER     = UINT32_MAX / 8,
    DEFAULT_TCP_ACK_NUMBER = 0x12390,
    MAX_TCP_ACK_NUMBER     = UINT32_MAX / 8,
    DEFAULT_TCP_FLAGS      = ACK_FLAG,
    DEFAULT_WND_SIZE       = 8192,
    MIN_VLAN_ID            = 1,
    MAX_VLAN_ID            = 4095,
    DEFAULT_VLAN_ID        = MIN_VLAN_ID,
    MIN_COS                = 0,
    MAX_COS                = 7,
    DEFAULT_COS            = MIN_COS,
    MIN_TOS                = 0,
    MAX_TOS                = 255,
    DEFAULT_TOS            = MIN_TOS,
    MAX_ETHER_TYPE_SIZE    = 0x600,
    OVERHEAD_FUDGE_VALUE   = 50,

    DEFAULT_PORTS_PER_PAGE = 4,
    VLAN_TAG_SIZE          = 4,
    MAX_PRIME_COUNT        = 4,

    NUM_SEQ_PKTS = 16, /* Number of buffers to support in sequence */

    FIRST_SEQ_PKT  = 0,
    SINGLE_PKT     = (FIRST_SEQ_PKT + NUM_SEQ_PKTS), /* 16 */
    PING_PKT       = (SINGLE_PKT + 1),               /* 17 */
    RANGE_PKT      = (PING_PKT + 1),                 /* 18 */
    DUMP_PKT       = (RANGE_PKT + 1),                /* 19 */
    RATE_PKT       = (DUMP_PKT + 1),                 /* 20 */
    LATENCY_PKT    = (RATE_PKT + 1),                 /* 21 */
    NUM_TOTAL_PKTS = (LATENCY_PKT + 1),

    INTER_FRAME_GAP       = 12, /**< in bytes */
    START_FRAME_DELIMITER = 1,  /**< Starting frame delimiter bytes */
    PKT_PREAMBLE_SIZE     = 7,  /**< in bytes */
    /* total number of bytes in frame overhead, includes the FCS checksum byte count */
    PKT_OVERHEAD_SIZE =
        (INTER_FRAME_GAP + START_FRAME_DELIMITER + PKT_PREAMBLE_SIZE + RTE_ETHER_CRC_LEN),
    MIN_v6_PKT_SIZE = 78, /**< does include FCS bytes */

    MAX_RX_QUEUES  = 16, /**< RX Queues per port */
    MAX_TX_QUEUES  = 16, /**< TX Queues per port */
    PCAP_PAGE_SIZE = 25  /**< Size of the PCAP display page */
};

#define WIRE_SIZE(pkt_size, t) (t)(pkt_size + PKT_OVERHEAD_SIZE)

typedef struct rte_mbuf rte_mbuf_t;

typedef union {
    struct rte_ether_addr addr;
    uint64_t u64;
} ethaddr_t;

/* Ethernet addresses of ports */
typedef struct pktgen_s {
    int verbose;                           /**< Verbose flag */
    int32_t argc;                          /**< Number of arguments */
    uint32_t blinklist;                    /**< Port list for blinking the led */
    uint32_t flags;                        /**< Flag values */
    volatile int force_quit;               /**< Flag to force quit */
    struct cmdline *cl;                    /**< Command Line information pointer */
    char *argv[64];                        /**< Argument list */
    char *hostname;                        /**< GUI hostname */
    int32_t socket_port;                   /**< GUI port number */
    volatile uint8_t is_gui_running;       /**< flag to denote GUI is running */
    volatile uint8_t timer_running;        /**< flag to denote timer is running */
    uint16_t ident;                        /**< IPv4 ident value */
    uint16_t last_row;                     /**< last static row of the screen */
    uint16_t nb_ports;                     /**< Number of ports in the system */
    uint8_t starting_port;                 /**< Starting port to display */
    uint8_t ending_port;                   /**< Ending port to display */
    uint8_t nb_ports_per_page;             /**< Number of ports to display per page */
    uint16_t mbuf_dataroom;                /**< Size of data room in mbuf */
    uint16_t mbuf_buf_size;                /**< MBUF default buf size */
    uint16_t nb_rxd;                       /**< Number of receive descriptors */
    uint16_t nb_txd;                       /**< Number of transmit descriptors */
    uint16_t curr_port;                    /**< Current Port number */
    uint64_t hz;                           /**< Number of cycles per seconds */
    uint64_t page_timeout;                 /**< Timeout for page update */
    uint64_t stats_timeout;                /**< Timeout for stats update */
    uint64_t max_total_ipackets;           /**< Total Max seen input packet rate */
    uint64_t max_total_opackets;           /**< Total Max seen output packet rate */
    uint64_t counter;                      /**< A debug counter */
    uint64_t mem_used;                     /**< Display memory used counters per ports */
    uint64_t total_mem_used;               /**< Display memory used for all ports */
    struct rte_eth_stats cumm_rate_totals; /**< port rates total values */

#ifdef LUA_ENABLED
    luaData_t *ld;      /**< General Lua Data pointer */
    luaData_t *ld_sock; /**< Info for Lua Socket */
    pthread_t thread;   /**< Thread structure for Lua server */
#endif

    uint8_t *portdesc[RTE_MAX_ETHPORTS];   /**< Port descriptions from lspci */
    uint32_t portdesc_cnt;                 /**< Number of ports in portdesc array */
    lscpu_t *lscpu;                        /**< CPU information */
    capture_t capture[RTE_MAX_NUMA_NODES]; /**< Packet capture, 1 struct per socket */
} pktgen_t;

enum {                                     /* Pktgen flags bits */
       PRINT_LABELS_FLAG      = (1 << 0),  /**< Print constant labels on stats display */
       MAC_FROM_ARP_FLAG      = (1 << 1),  /**< Configure the SRC MAC from a ARP request */
       PROMISCUOUS_ON_FLAG    = (1 << 2),  /**< Enable promiscuous mode */
       NUMA_SUPPORT_FLAG      = (1 << 3),  /**< Enable NUMA support */
       IS_SERVER_FLAG         = (1 << 4),  /**< Pktgen is a Server */
       ENABLE_GUI_FLAG        = (1 << 5),  /**< GUI support is enabled */
       LUA_SHELL_FLAG         = (1 << 6),  /**< Enable Lua Shell */
       TX_DEBUG_FLAG          = (1 << 7),  /**< TX Debug output */
       Not_USED               = (1 << 8),  /**< Not Used */
       FAKE_PORTS_FLAG        = (1 << 9),  /**< Fake ports enabled */
       BLINK_PORTS_FLAG       = (1 << 10), /**< Blink the port leds */
       ENABLE_THEME_FLAG      = (1 << 11), /**< Enable theme or color support */
       CLOCK_GETTIME_FLAG     = (1 << 12), /**< Enable clock_gettime() instead of rdtsc() */
       JUMBO_PKTS_FLAG        = (1 << 13), /**< Enable Jumbo frames */
       RESERVED_14            = (1 << 14),
       RESERVED_15            = (1 << 15),
       CPU_PAGE_FLAG          = (1 << 16), /**< Display the CPU page */
       SEQUENCE_PAGE_FLAG     = (1 << 17), /**< Display the Packet sequence page */
       RANGE_PAGE_FLAG        = (1 << 18), /**< Display the range page */
       PCAP_PAGE_FLAG         = (1 << 19), /**< Display the PCAP page */
       SYSTEM_PAGE_FLAG       = (1 << 20), /**< Display the System page */
       RND_BITFIELD_PAGE_FLAG = (1 << 21), /**< Display the random bitfield page */
       LOG_PAGE_FLAG          = (1 << 22), /**< Display the message log page */
       LATENCY_PAGE_FLAG      = (1 << 23), /**< Display latency page */
       STATS_PAGE_FLAG        = (1 << 24), /**< Display the physical port stats */
       XSTATS_PAGE_FLAG       = (1 << 25), /**< Display the physical port stats */
       RATE_PAGE_FLAG         = (1 << 26), /**< Display the Rate Pacing stats */
       RESERVED_27            = (1 << 27),
       RESERVED_28            = (1 << 28),
       RESERVED_29            = (1 << 29),
       RESERVED_30            = (1 << 30),
       UPDATE_DISPLAY_FLAG    = (1 << 31)
};

#define UPDATE_DISPLAY_TICK_INTERVAL 4 /* check stats rate per second */
#define UPDATE_DISPLAY_TICK_RATE     (pktgen.hz / UPDATE_DISPLAY_TICK_INTERVAL)

#define PAGE_MASK_BITS                                                                          \
    (CPU_PAGE_FLAG | SEQUENCE_PAGE_FLAG | RANGE_PAGE_FLAG | PCAP_PAGE_FLAG | SYSTEM_PAGE_FLAG | \
     RND_BITFIELD_PAGE_FLAG | LOG_PAGE_FLAG | LATENCY_PAGE_FLAG | XSTATS_PAGE_FLAG |            \
     STATS_PAGE_FLAG | RATE_PAGE_FLAG)

extern pktgen_t pktgen;

void pktgen_page_display(void);

void pktgen_packet_ctor(port_info_t *pinfo, int32_t seq_idx, int32_t type);
void pktgen_packet_rate(port_info_t *pinfo);

int pktgen_find_matching_ipsrc(port_info_t *pinfo, uint32_t addr);
int pktgen_find_matching_ipdst(port_info_t *pinfo, uint32_t addr);

int pktgen_launch_one_lcore(void *arg);
void pktgen_input_start(void);
void stat_timer_dump(void);
void stat_timer_clear(void);
void pktgen_timer_setup(void);
double next_poisson_time(double rateParameter);

static inline uint64_t
pktgen_get_time(void)
{
    if (pktgen.flags & CLOCK_GETTIME_FLAG) {
        struct timespec tp;

        if (clock_gettime(CLOCK_REALTIME, &tp) < 0)
            return rte_rdtsc_precise();

        return rte_timespec_to_ns(&tp);
    } else
        return rte_rdtsc_precise();
}

static inline uint64_t
pktgen_get_timer_hz(void)
{
    if (pktgen.flags & CLOCK_GETTIME_FLAG) {
        struct timespec tp = {.tv_nsec = 0, .tv_sec = 1};
        return rte_timespec_to_ns(&tp);
    } else
        return rte_get_timer_hz();
}

typedef struct {
    uint32_t magic;
    uint32_t index;
    uint64_t timestamp;
} tstamp_t;

#define TSTAMP_MAGIC 0xf00dcafe

static __inline__ void
pktgen_set_port_flags(port_info_t *pinfo, uint32_t flags)
{
    uint32_t val;

    do {
        val = rte_atomic32_read(&pinfo->port_flags);
    } while (!rte_atomic32_cmpset((volatile uint32_t *)&pinfo->port_flags.cnt, val, (val | flags)));
}

static __inline__ void
pktgen_clr_port_flags(port_info_t *pinfo, uint32_t flags)
{
    uint32_t val;

    do {
        val = rte_atomic32_read(&pinfo->port_flags);
    } while (
        !rte_atomic32_cmpset((volatile uint32_t *)&pinfo->port_flags.cnt, val, (val & ~flags)));
}

static __inline__ int
pktgen_tst_port_flags(port_info_t *pinfo, uint32_t flags)
{
    return (rte_atomic32_read(&pinfo->port_flags) & flags) ? 1 : 0;
}

/* onOff values */
enum { DISABLE_STATE = 0, ENABLE_STATE = 1 };

static __inline__ uint32_t
estate(const char *state)
{
    return (!strcasecmp(state, "on") || !strcasecmp(state, "enable") || !strcasecmp(state, "start"))
               ? ENABLE_STATE
               : DISABLE_STATE;
}

/* LatSampler types */
enum { LATSAMPLER_UNSPEC, LATSAMPLER_SIMPLE, LATSAMPLER_POISSON };

/**
 * Function returning string of version number: "- Version:x.y.x (DPDK-x.y.z)"
 * @return
 *     string
 */
static inline const char *
pktgen_version(void)
{
    static char pkt_version[64];

    if (pkt_version[0] != 0)
        return pkt_version;

    snprintf(pkt_version, sizeof(pkt_version), "%s (%s)", PKTGEN_VERSION, rte_version());
    return pkt_version;
}

static __inline__ char *
strdupf(char *str, const char *new)
{
    if (str)
        free(str);
    return (new == NULL) ? NULL : strdup(new);
}

/**
 *
 * do_command - Internal function to execute a shell command and grab the output.
 *
 * DESCRIPTION
 * Internal function to execute a shell command and grab the output from the
 * command.
 *
 * RETURNS: Nubmer of lines read.
 *
 * SEE ALSO:
 */

static __inline__ int
do_command(const char *cmd, int (*display)(char *, int))
{
    FILE *f;
    int i;
    char *line       = NULL;
    size_t line_size = 0;

    f = popen(cmd, "r");
    if (f == NULL) {
        pktgen_log_error("Unable to run '%s' command", cmd);
        return -1;
    }

    i = 0;
    while (getline(&line, &line_size, f) > 0)
        i = display(line, i);

    if (f)
        pclose(f);
    if (line)
        free(line);

    return i;
}

#ifndef MEMPOOL_F_DMA
#define MEMPOOL_F_DMA 0
#endif

#ifdef __cplusplus
}
#endif

#endif /* _PKTGEN_H_ */
